Sprint 17

Dimaulta opened the pull request:
Сдача приложения Tracker по чек-листу 17-го спринта

Dimaulta committed:
Инициалищация 17-го спринта

Dimaulta committed:
feat: Написал механизм редактирование и удаление трекеров

- Добавил инициализатор Tracker с id для сохранения идентичности объектов
- Исправил маппинг из Core Data: используем сохранённый id вместо генерации
- Улучшил обновление трекера: сохраняются имя, emoji, цвет и расписание
- Добавил обновление категории при редактировании, связь корректно пересохраняется

Dimaulta committed:
feat: добавил локализацию RU и плюрализацию дней

- Добавил файлы локализации Localizable.strings и Localizable.stringsdict (ru.lproj)
- Перенёс строковые литералы в Localizable.strings
- Заменил все строки в UI на NSLocalizedString
- Улучшил отображение «N дней» с корректными окончаниями через stringsdict
- Оставил поддержку только русского языка

Dimaulta committed:
feat: добавил экран и кнопку фильтров

- Создал экран выбора фильтра для привычек
- Добавил локализационные ключи для пунктов фильтра
- Добавил библиотеку snapshot тестов (тесты не написаны)

Dimaulta committed:
feat: добавил поиск по всем трекерам вне зависимости от дня

- Если поле поиска пустое — показываю трекеры за выбранный день
- Если есть запрос — ищу по всем трекерам, затем применяю выбранный фильтр

Dimaulta committed:
feat: добавил фильтрацию привычек с учетом выбранной даты

- Реализовал фильтры по требованию чеклиста
- Добавил кнопку "Фильтры" на главный экран с автоматическим скрытием при отсутствии трекеров на выбранный день
- Настроил отображение галочек только для фильтров "Завершённые" и "Незавершённые"
- Добавил автоматическое закрытие экрана фильтров после выбора

Dimaulta committed:
feat: Добавил английскую локализацию для главного экрана

- Добавил файл английской локализации en.lproj/Localizable.strings
- Локализовал заголовок "Трекеры" и кнопку "Фильтры"
- Локализовал названия табов "Трекеры" и "Статистика"
- Обновил TabBarController для использования NSLocalizedString

Dimaulta committed:
fix: Исправил плюрализацию для дней в локализации

- Создал английский .stringsdict файл
- Исправил формы слов в .strings файлах
- Теперь правильно отображается "1 day" и "1 день"

Dimaulta committed:
feat: Добавил snapshot тесты для главного экрана

- Интегрировал библиотеку SnapshotTesting (ранее)
- Создал тесты для пустого состояния и состояния с данными трекеров
- Добавил MockTrackerViewModel для изоляции тестов от реальных данных
- Создал эталонные скриншоты для автоматической проверки UI (по мок данным)

Dimaulta committed:
feat: Создал экран статистики с пустым состоянием

- Добавил StatisticsViewController с заголовком "Статистика"
- Добавил локализацию на русский и английский языки
- Пока без привязки к реальным данным (заглушка с api ключом в коде)

Dimaulta committed:
feat: Добавил адаптацию под тёмную тему

- Заменил все прямые RGB цвета из ассетов
- Унифицировал фоны экранов и ячеек для всех экранов приложения
- Исправил цвета текста кнопок для корректного отображения в тёмной теме
- Настроил цвета плюсиков трекеров для адаптации к тёмной теме

todo: написать статистику по подключенной SDK Яндекс аналитика

Dimaulta committed:
feat: Добавил локальную статистику и аналитику YandexMobileMetrica (аналитика не реализована)

- Создал StatisticsManager для вычисления реальных метрик из CoreData
- Реализовал расчет лучшего периода, идеальных дней, завершенных трекеров и среднего значения
- Интегрировал StatisticsManager в StatisticsViewController с автообновлением
- Добавил AnalyticsManager для отправки событий в YandexMobileMetrica (аналитика не реализована)
- Настроил отслеживание создания/редактирования трекеров и категорий
- Добавил события завершения трекеров, просмотра статистики и использования фильтров
- Создал градиентные рамки карточек с переходом красный-зеленый-синий из ассетов

todo: реализовать аналитику

Dimaulta committed:
feat: Добавил аналитику YandexMobileMetrica согласно требованиям учебника

- Настроил SDK с API ключом (ключ учебного приложения)
- Реализовал отправку событий открытия/закрытия экрана Main
- Добавил аналитику тапов на кнопки (add_track, track, filter, edit, delete)
- Создал систему тестирования с подробным логированием
- Добавил документацию и чек-лист для проверки работы аналитики

Helvetii951 commented on Tracker/AnalyticsManager.swift:
```diff
@@ -0,0 +1,220 @@
+    
+    // MARK: - AppMetrica Events (согласно требованиям учебника)
+    
+    func trackScreenOpen(screen: String) {
+        let parameters: [String: Any] = [
+            "event": "open",
+            "screen": screen
+        ]
+        
+        YMMYandexMetrica.reportEvent("event", parameters: parameters)
+        
+        #if DEBUG
+        print("📊 Отправлено событие: screen_open - \(screen)")
+        #endif
+    }
+    
+    func trackScreenClose(screen: String) {
+        let parameters: [String: Any] = [
+            "event": "close",
+            "screen": screen
+        ]
+        
+        YMMYandexMetrica.reportEvent("event", parameters: parameters)
+        
+        #if DEBUG
+        print("📊 Отправлено событие: screen_close - \(screen)")
+        #endif
+    }
+    
+    func trackButtonClick(screen: String, item: String) {
+        let parameters: [String: Any] = [
```
Функции лучше сделать статическими, тогда нам не будет нужен экземпляр `shared`.

Helvetii951 commented on Tracker/AppDelegate.swift:
```diff
@@ -7,13 +7,39 @@
 
-
-
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+     
+        let apiKey = "2f9bfdfc-6406-4df5-a776-e7d876db84cd"
+        
+        if let configuration = YMMYandexMetricaConfiguration(apiKey: apiKey) {
+         
+            #if DEBUG
+            configuration.logs = true
+            configuration.crashReporting = true
+            #endif
+            
+            YMMYandexMetrica.activate(with: configuration)
```
Активацию лучше вынести в обертку `AnalyticsManager`.

Helvetii951 commented on Tracker/CategoryViewController.swift:
```diff
@@ -87,8 +87,8 @@ final class CategoryViewController: UIViewController {
-        addCategoryButton.setTitle("Добавить категорию", for: .normal)
-        addCategoryButton.setTitleColor(UIColor.white, for: .normal)
+        addCategoryButton.setTitle(NSLocalizedString("category.add.button", comment: "Добавить категорию"), for: .normal)
+        addCategoryButton.setTitleColor(UIColor(named: "WhiteNight"), for: .normal)
```
Для получения цвета надежнее будет воспользоваться генерацией ресурсов xCode 15+:
```swift
addCategoryButton.setTitleColor(UIColor(resource: .whitenight), for: .normal)
```

Так мы будем уверены, что цвет в проекте есть и не нужно будет применять nil coalesing(`??`).

Подробнее можно почитать [здесь](https://sarunw.com/posts/swift-symbols-for-asset-catalog/).

Helvetii951 commented on Tracker/FiltersViewController.swift:
```diff
@@ -0,0 +1,206 @@
+    
+    var title: String {
+        switch self {
+        case .all: return NSLocalizedString("filters.all", comment: "Все трекеры")
+        case .today: return NSLocalizedString("filters.today", comment: "Трекеры на сегодня")
+        case .completed: return NSLocalizedString("filters.completed", comment: "Завершенные")
+        case .incomplete: return NSLocalizedString("filters.incomplete", comment: "Не завершенные")
```
Советую обратить внимание на еще два способа работы с локализованными строками:

1) Использовать библиотеки [R.Swift ](https://github.com/mac-cain13/R.swift)или [SwiftGen](https://github.com/SwiftGen/SwiftGen).
2) Создать String Catalog [с помощью xCode](https://developer.apple.com/documentation/xcode/localizing-and-varying-text-with-a-string-catalog).

Helvetii951 commented on Tracker/FiltersViewController.swift:
```diff
@@ -0,0 +1,206 @@
+        
+        titleLabel.translatesAutoresizingMaskIntoConstraints = false
+        titleLabel.text = NSLocalizedString("filters.title", comment: "Фильтры")
+        titleLabel.font = UIFont(name: "SFPro-Medium", size: 16) ?? UIFont.systemFont(ofSize: 16, weight: .medium)
```
SFPro - системный шрифт в iOS. Лучше всегда пользоваться [systemFont(ofSize:weight:)](https://developer.apple.com/documentation/uikit/uifont/1619027-systemfont).

Helvetii951 commented on Tracker/FiltersViewController.swift:
```diff
@@ -0,0 +1,206 @@
+final class FiltersViewController: UIViewController {
+    
+    // MARK: - UI
+    private let titleLabel = UILabel()
+    private let containerView = UIView()
+    private let tableView = UITableView()
+    
```
Рекомендую всегда использовать [ленивые свойства](https://www.swiftbysundell.com/articles/using-lazy-properties-in-swift/) для объявления подобного.  В нашем случае это какого-то улучшения производительности не даст. Но, если какой-то UI компонент/свойство не будет использоваться, можно сэкономить память. 

Подход с ленивыми свойствами используют во многих хороших компаниях

Helvetii951 commented on TrackerTests/__Snapshots__/TrackersViewControllerSnapshotTests/testTrackersViewControllerEmptyState.1.png:
Лучше не хранить сгенеренные сторонними библиотеками файлы в репозитории, а выносить в .gitignore.

Dimaulta committed:
feat: Исправил логику выполнения трекеров

- Убрал ограничение на выполнение трекеров в будущих днях
- Теперь трекеры можно выполнять в любой день, если они запланированы на этот день

Dimaulta committed:
feat: создал контекстное меню вместо actionSheet

- Проработал меню

Dimaulta committed:
feat: контекстное меню по макету

- Теперь контекстное меню соответствует макету figma

Dimaulta committed:
feat: Алерт подтверждения удаления привычки

- Создал отдельный файл для кастомного алерта
- Начал подгонять его под макет figma

Dimaulta committed:
feat: Алерт подтверждения удаления привычки

- Полностью по макету figma

Dimaulta committed:
feat: Улучшил UI консистентность и добавил блокировку выполнения привычек в будущих датах

- Заменил кастомный алерт удаления на стандартный UIAlertController с actionSheet
- Исправил цвета текста дней недели и кнопки "Готово" в расписании на цвета из ассетов
- Добавил блокировку выполнения привычек в будущих датах с анимацией покачивания
- Унифицировал цвета плюсика и галочки в трекерах
- Настроил динамическое изменение состояния кнопки "Готово" в расписании

Dimaulta committed:
feat: Исправил layout главного экрана и кнопки Фильтры

- Сократил расстояние между полем поиска и заголовком категории с 34 до 1 пункта
- Убрал избыточное белое пространство под полем поиска
- Исправил позиционирование кнопки "Фильтры" - теперь она поверх списка привычек
- Убрал белую плашку вокруг кнопки "Фильтры", теперь цветные ячейки видны по бокам
- Сделал коллекцию полноэкранной с contentInset для корректного скролла

Dimaulta committed:
feat: Исправил неработающую кнопку фильтров

- Обнаружил проблему с кнопкой фильтров - она не реагировала на нажатия
- Выяснил причину: UICollectionView перекрывал кнопку фильтров
- Изменил constraints CollectionView - теперь он заканчивается на 120px выше низа экрана
- Переместил кнопку фильтров в правильное место - внизу экрана, но выше TabBar
- Изменил порядок добавления view - кнопка теперь добавляется после CollectionView
- Добавил bringSubviewToFront для явного поднятия кнопки наверх
- Убрал отладочный код и тестовые элементы
- Кнопка фильтров теперь корректно работает и открывает экран фильтров
